evennia/contrib/cooldowns.py

```diff
@@ -0,0 +1,188 @@
+
+    Methods:
+    - ready(name): Checks whether a given cooldown name is ready.
+    - time_left(name): Returns how much time is left on a cooldown.
+    - set(name, seconds): Sets a given cooldown to last for a certain
+        amount of time. Until then, ready() will return False for that
+        cooldown name.
+    - extend(name, seconds): Like set, but adds time to the given
+        cooldown name. If it doesn't exist yet, calling this is equivalent
+        to calling set.
+    - reset(cooldown): Resets a given cooldown, causing ready() to return
+        True for that cooldown immediately.
+    - clear(): Resets all cooldowns.
+    """
+
+    def __init__(self, obj, db_attribute="cooldowns"):
+        if not obj.attributes.has(db_attribute):
+            obj.attributes.add(db_attribute, {})
+
+        self.data = obj.attributes.get(db_attribute)
+        self.cleanup()
+
+    @property
+    def all(self):
+        """
+        Returns a list of all keys in this object.
+        """
+        return list(self.data.keys())
+
+    def ready(self, *args):
+        """
+        Checks whether all of the provided cooldowns are ready (expired).
+        If a requested cooldown does not exist, it is considered ready.
+
+        Args:
+            any (str): One or more cooldown names to check.
+        Returns:
+            (bool): True if each cooldown has expired or does not exist.
+        """
+        return self.time_left(*args) <= 0
+
+    def time_left(self, *args):
+        """
+        Returns the maximum amount of time left on one or more given
+        cooldowns. If a requested cooldown does not exist, it is
+        considered to have 0 time left.
+
+        Args:
+            any (str): One or more cooldown names to check.
+        Returns:
+            (int): Number of seconds until all provided cooldowns are
+                ready. Returns 0 if all cooldowns are ready (or don't
+                exist.)
+        """
+        now = time.time()
+        cooldowns = [self.data[x] - now for x in args if x in self.data]
+        if not cooldowns:
+            return 0
+        return math.ceil(max(max(cooldowns), 0))
+
+    def set(self, cooldown, seconds):
+        """
+        Sets a given cooldown to last for a specific amount of time.
+
+        If this cooldown is already set, this replaces it.
+
+        Args:
+            cooldown (str): The name of the cooldown.
+            seconds (int): The number of seconds before this cooldown is
+                ready again.
+        """
+        now = time.time()
+        self.data[cooldown] = int(now) + (max(seconds, 0) if seconds else 0)
+
+    def extend(self, cooldown, seconds):
+        """
+        Adds a specific amount of time to an existing cooldown.
+
+        If this cooldown is already ready, this is equivalent to calling
+        set. If the cooldown is not ready, it will be extended by the
+        provided duration.
```

Comment:
It feels like the outcome of this operation can vary quite a bit. It's basically either an extend or an add/set; the result of which depends on what existed beforehand. You may not always know this when calling, and could end up with a longer delay than you anticipated. How about returning the resulting time-left for calling code that cares?

Reply:
Good point. Added a return value for time left.